cudax/include/cuda/experimental/__utility/basic_any/semiregular.cuh

```diff
@@ -145,37 +149,108 @@ struct icopyable : interface<icopyable, extends<imovable<>>>
-  _CCCL_NODISCARD _CUDAX_HOST_API auto operator==(iequality_comparable const& __other) const -> bool
+  _CCCL_TEMPLATE(class _ILeft, class _IRight)
+  _CCCL_REQUIRES(__any_convertible_to<basic_any<_ILeft> const&, basic_any<_IRight> const&>
+                 || __any_convertible_to<basic_any<_IRight> const&, basic_any<_ILeft> const&>)
+  _CCCL_NODISCARD_FRIEND _CUDAX_TRIVIAL_HOST_API auto
+  operator!=(iequality_comparable<_ILeft> const& __lhs, iequality_comparable<_IRight> const& __rhs) noexcept -> bool
+  {
+    return !(__lhs == __rhs);
+  }
+
+  // These are the overloads that get used when testing a `basic_any` object
+  // against a non-type-erased object.
+  //
+  // Q: Why require that the basic_any wrapper is not convertible to _Object?
+  //
+  // A: If there is a user-defined conversion from the basic_any type to _Object
+  // such that we can use _Object's symmetric equality comparison operator, that
+  // should be prefered. _Object may be another kind of wrapper object, in which
+  // case using the address of the **wrapper** object for the comparison (as
+  // opposed to the address of the wrapped object) is probably wrong.
+  _CCCL_TEMPLATE(class _Interface, class _Object, class _Self = basic_any_from_t<iequality_comparable<_Interface>>)
+  _CCCL_REQUIRES(__non_polymorphic<_Object> _CCCL_AND(!_CUDA_VSTD::convertible_to<_Self, _Object>)
+                   _CCCL_AND __satisfies<_Object, _Interface>)
+  _CCCL_NODISCARD_FRIEND _CUDAX_HOST_API auto
+  operator==(iequality_comparable<_Interface> const& __lhs, _Object const& __rhs) -> bool
   {
-    auto const& __rhs = __cudax::basic_any_from(__other);
-    void const* __obj = __basic_any_access::__get_optr(__rhs);
-    return __cudax::virtcall<&__equal_fn<iequality_comparable>>(this, __rhs.type(), __obj);
+    constexpr auto __eq = &__equal_fn<iequality_comparable<_Interface>>;
+    return __cudax::virtcall<__eq>(&__lhs, _CCCL_TYPEID(_Object), _CUDA_VSTD::addressof(__rhs));
   }
-#else // ^^^ !_CCCL_NO_THREE_WAY_COMPARISON ^^^ / vvv _CCCL_NO_THREE_WAY_COMPARISON vvv
+
+  _CCCL_TEMPLATE(class _Interface, class _Object, class _Self = basic_any_from_t<iequality_comparable<_Interface>>)
+  _CCCL_REQUIRES(__non_polymorphic<_Object> _CCCL_AND(!_CUDA_VSTD::convertible_to<_Self, _Object>)
+                   _CCCL_AND __satisfies<_Object, _Interface>)
   _CCCL_NODISCARD_FRIEND _CUDAX_HOST_API auto
-  operator==(iequality_comparable const& __left, iequality_comparable const& __right) -> bool
+  operator==(_Object const& __lhs, iequality_comparable<_Interface> const& __rhs) noexcept -> bool
   {
-    auto const& __rhs = __cudax::basic_any_from(__right);
-    void const* __obj = __basic_any_access::__get_optr(__rhs);
-    return __cudax::virtcall<&__equal_fn<iequality_comparable>>(&__left, __rhs.type(), __obj);
+    constexpr auto __eq = &__equal_fn<iequality_comparable<_Interface>>;
+    return __cudax::virtcall<__eq>(&__rhs, _CCCL_TYPEID(_Object), _CUDA_VSTD::addressof(__lhs));
   }
```

Comment:
NVCC has notorious issues when constraining friend functions. I would be wary of that formulation and use `_CCCL_TRAILING_REQUIRES(bool)(...)`